spec-0012/index.md

```diff
@@ -4,122 +4,201 @@ number: 12
+regarded as `(...)`.
+- `x * y + z` is a compound expression; there are two operators and no explicit
+subexpressions that can be ignored.
+
+The acronym PEMDAS commonly refers to "parentheses", "exponentiation", "multiplication",
+"division", "addition", and "subtraction". Herein, we will consider these operators
+to be "PEMDAS operators", and we will also include the unary `+`, `-`, and `~` in
+this category for convenience. The order of operations of PEMDAS operators is typically
+taught in primary school and reinforced throughout a programmer's training and
+experience, so it is assumed that most programmers are comfortable relying on the
+implicit order of operations of expressions involving a few PEMDAS operations. Implicit
+order of operations becomes less obvious as the number of distinct operator priority
+levels increases and when multiple non-PEMDAS operators are involved. Portions of this
+acronym, namely MD and AS, will be used below to refer to the corresponding operators.
 
-## Notes
+## Implementation
 
-These formatting rules do not make any consideration in terms of performances
-nor precision. The scope is limited to styling.
+The following rules are not imagined to be all encompassing, and there may be
+syntactically valid cases in which the following rules would lead to contradictions.
+Nonetheless, we expect them to be useful in most cases, and projects that adopt the
+SPEC endeavor to use them where practical. These rules are intended to respect and
+complement the [PEP 8 standards](https://peps.python.org/pep-0008), such as using
+[implied line continuation](https://peps.python.org/pep-0008/#maximum-line-length) and
+and [breaking lines before binary operators](https://peps.python.org/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).
+
+0. Unless otherwise specified, rely on the implicit order of operations;
+   i.e., do not add extraneous parentheses. For example, prefer `u**v + y**z`
+   over `(u**v) + (y**z)`, and prefer `x + y + z` over `(x + y) + z`. A full
+   list of implicit operator priority levels is given by
+   [Operator Precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence)
+1. Always use the `**` operator and unary `+`, `-`, and `~` operators *without*
+   surrounding whitespace. For example, prefer `-x**4` over `- (x ** 4)`. 
+2. Always surround non-PEMDAS operators with whitespace, and always make the priority of
+   non-PEMDAS operators explicit. For example, prefer `(x == y) or (w == t)` over
+   `x==y or w==t`.[^1]
+3. Always surround AS operators with whitespace.
+4. Typically, surround MD operators with whitespace, except in the following situations.
+   - When there are lower-priority operators (namely AS) within the same compound
+     expression. For example, prefer `z = -x * y**t` over `z = -x*y**t`, but
+     prefer `z = w + x*y**t` over `z = w + x * y**t` due to the presence of the
```

Comment:
As below, I think there should be parentheses here to distinguish between `*` and `**`:

`z = w + x*(y**t)`

Reply:
Also, how would the rules format `w * x * y**z + u ** v * q`?

I think my preference is:

`(w * x * y**z) + (u**v * q)`

But I would accept:

`(w*x*(y**z)) + ((u**v)*q)`

Reply:
Here, the presence of the lower priority operator `+` in the expression means that there will be no space around the multiplication operators. Also, `**` is supposed to appear at the end of "implicit subexpression without spaces". So it would be `w*x*y**z + q*u**v`.

This is intended to be reminiscent of mathematics in which we would write
$wxy^z + qu^v$.

Reply:
What about `w*x*(y**z) + q*(u**v)`? I would be happy with that.

Reply:
We need a stronger rationale than what makes one of us happy to change that : ) 

Would you really write $xy^z$ as $x(y^z)$?

And you'll say no, but the superscript stands out so it's readable. Then I say ok, well I think the `**` stands out so it's readable.

Is what you're suggesting much more common in code already? That would be a good reason to change.

Reply:
I'm not sure what's more common, but indeed the problem for me is that it doesn't stand out. Especially for longer expressions such as

`a + b*c*d*e**f`

so we just disagree on

> I think putting ** at the end of subexpressions without spaces is enough to distinguish it visually from *.

But yeah, I'm happy to be outvoted on this one. Just for the record!

Reply:
And yeah, I think that's what we need is more opinions. Maybe we could add a limit on the number of chained operators in subexpressions without spaces, because I agree that `b*c*d*e**f` is not great style.